### ISSUE TITLE
Compare / Contrast forum module options

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/forum.backdropcms.org#3]
Let's compare three different forum solutions to see what people like / dislike about each:
1) Content type + Taxonomy + Comments
2) [Forum](https://github.com/backdrop-contrib/forum) module
3) [QA Forum](https://github.com/serundeputy/qa_forum) module

I'll set up [forum.backdropcms.org](https://forum.backdropcms.org) where we can evaluate one of each of these forums, and see what we think!

**Comment by Dinornis** [comment]
Just wondering, would it be beneficial to start this process with identifying a list of features needed for b.o forum?

Does b.org need a QA type forum, discussion type forum or both?

QA type forum can probably be built using the QA Forum module by serundeputy or using Backdrop core features + contrib modules.

Discussion type forum could be built with the contrib Forum module + number of other contrib modules.  It all depends on the features needed.

Seems like a quite functional discussion type forum could be put together with a few already existing D7 modules?

**Comment by jenlampton** [comment]
> Just wondering, would it be beneficial to start this process with identifying a list of features needed for b.o forum?

We're way ahead of you.  See [this old issue](https://github.com/backdrop-ops/backdropcms.org/issues/250) now moved into this [new meta](https://github.com/backdrop-ops/forum.backdropcms.org/issues/4)

> Does b.org need a QA type forum, discussion type forum or both?

both.

> Seems like a quite functional discussion type forum could be put together with a few already existing D7 modules?

Yes, or better yet, with backdrop modules!

**Comment by Dinornis** [comment]
Great, thanks.
Will the video from the forum meeting or notes from it available?

**Comment by jenlampton** [comment]
There was no video recorded of the meeting, and the notes are what's in the issue here: https://github.com/backdrop-ops/backdropcms.org/issues/250 - both the list at the top as well as my 2nd to last comment at the bottom. 

I think right now we're blocked on technical direction - this issue. We're likely to choose 1) or 2) from the first post in this issue, but it would be good if we all got a little hands-on side-by-side comparison experience before we make that decision :)

I've *almost* got the server working at  [forum.backdropcms.org](https://forum.backdropcms.org) but I'm not quite as familair with NGINX as I am with Apache, so have hit a snag. Time to ask @quicksketch for some guidance...

**Comment by tomgrandy** [comment]
Thanks for getting this setup @jenlampton !

**Comment by docwilmot** [comment]
Just to complicate matters as much as possible:
- [Brain forum](https://github.com/backdrop-contrib/brain_forum) is a port of a module which attempts to build forums without the dependency on nodes and comments. @serundeputy and I had a brief discussion on Gitter where I mentioned that essentially QA forum is trying to get rid of the non-dependency, so this **could** be used as the skeleton for QA forum instead. I'm not sure but I suspect all the non-standard features we're looking for may be long run easier to accomplish without the baggage of nodes and comments (although @jenlampton will ask me to explain this and I wont be able to). 
- Also Forum module in contrib has a version 2 which uses Views instead of hardcoded lists; we may want to try that as well or instead.

In either case how do we proceed? I imagine we're not going to waste time building out all the [required features](https://github.com/backdrop-ops/forum.backdropcms.org/issues/5) until we chose one of the forum options, so how do we make that first decision? Are there tests we should attempt or something?

**Comment by jenlampton** [comment]
I've built most of the required features already, most of them don't depend
on that choice.

The choice is between forum and custom, and I'd like to see what the
difference is. What does forum module provide, and do we need it?

To determine this, I've set up both on the forum site. We should take them
for a spin!

If anyone has specific use-cases or user-stories to test, that would be
helpful :)

**Comment by tomgrandy** [comment]
Apparently credentials from backdrop.org do not carry over to forums.  Created a new account there, but am awaiting approval from the owner (assuming that is @jenlampton ).

**Comment by docwilmot** [comment]
> The choice is between forum and custom

In that case I go with custom. Sorting and adding additional bells and whistles like votes will be easier then.

> Apparently credentials from backdrop.org do not carry over to forums. 

Thats why [Bakery](https://github.com/backdrop-contrib/bakery) is needed.

**Comment by Dinornis** [comment]
By custom you mean custom contrib modules versus the contrib Forum module?
When the b.o forum is finished, will all the components (modules) be available to Backdrop users for use on their sites?

**Comment by serundeputy** [comment]
my intention with `qa_forum` was to flesh out the functionality we want for forum.backdropcms.org and release it in contrib.

but that is one persons idea.  I think it makes sense to release the custom as a new `contrib`.

**Comment by jenlampton** [comment]
Custom is just config. No custom or contrib code needed. We could bundle
and release the config as soon as we have recipes or features for backdrop.

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by Graham-72** [comment]
> No, Custom is just config. No custom or contrib code is needed.

I'm puzzled by this. Is custom not using the qa_forum module?

Incidentally I very much like your strapline 'Where all questions are answered with dignity and respect.'

Another point, I'm very much influenced by the appearance of listings of comments etc. This, which can be adjusted with Views and CSS, might prejudice my choice without me having any real understanding of the merits of the different undrlying technology. Put simply, if it look good and is easy to use I will like it!

**Comment by jenlampton** [comment]
> Is custom not using the qa_forum module?

No, "Forum topic - Question" is using the qa_forum module. "Forum topic - Custom" is not using any special module at all. "Forum topic - Forum" is using the forum module.

> I'm very much influenced by the appearance of listings of comments etc

Yes, this is understandable and likely applies to everyone. Right now all 3 options are using the same kinds of views and CSS so at least they are all on level ground for the review phase.

**Comment by docwilmot** [comment]
> Forum topic - Question

I dont see this. Maybe Anon permissions?

In either case, with all the "already implemented" labels youre suggesting @jenlampton, I'm suspecting most of this work is with the custom solution. I would think there is no reasonable way for us to "evaluate" the different forum solutions unless we actually attempt to build all the needed/wanted features with *each*, see how easy and/or successful we are with *each*, then compare notes and stats and see which worked out better.

But that is obviously unreasonable, so again, I'm voting to scrap the other two and just get down to business with the custom solution.

**Comment by jenlampton** [comment]
> I'm suspecting most of this work is with the custom solution.

No, they are all on all of the content types. Both with views and with flags, it is very easy to allow for more than one content type: just check another box. In this case, I checked three boxes. :)

> I would think there is no reasonable way for us to "evaluate" the different forum solutions unless we actually attempt to build all the needed/wanted features with each, see how easy and/or successful we are with each, then compare notes and stats and see which worked out better.

Yep, that's why they are all on all of them.

> I'm voting to scrap the other two and just get down to business with the custom solution.

We could do that, but I'd still be wondering: what does the forum module actually provide? If we didn't use it, what would we be loosing?

**Comment by Dinornis** [comment]
From user perspective view I already dislike the custom version, too confusing and dificult to navigate.

In the Forum (module) I am getting errors at https://forum.backdropcms.org/forum/2 and https://forum.backdropcms.org/forum/2#new

So far I like the Forum (module) option.

**Comment by Graham-72** [comment]
> Forum topic - Question" is using the qa_forum module.

I am not seeing a menu item for this, other than one to create a topic/question. Is there a URL for seeing a list of topics/questions as for the other two forums?

**Comment by Graham-72** [comment]
I am confused by the listing in Forum (Custom). Is it generating separate rows for individual contributors?
![forum custom backdrop forum](https://cloud.githubusercontent.com/assets/10559274/22415296/ec95c3f0-e6bd-11e6-93bf-9910f97f4519.png)
Forums (module) seems better.
![forums module backdrop forum](https://cloud.githubusercontent.com/assets/10559274/22415302/f83930fc-e6bd-11e6-8efe-e62735a17bb4.png)

**Comment by jenlampton** [comment]
It's likely a problem with how I set up aggregation on the view, will take a look.

**Comment by jenlampton** [comment]
>  Is it generating separate rows for individual contributors?

Yeah, it was doing that. For some reason it had different answers for "last comment author" even though it knew the "last comment time" was the same. For now I've removed authors, but we'll need to add them back (and fix that bug!) once we decide on a direction.

> I am not seeing a menu item for [qa_forum]

That's right. QA forum does not provide it's own page like forum module does (unless I am mistaken @serundeputy ?).  If we were to build one it would likely look exactly like forum (custom), so it's not part of the comparison equation. However, creating a topic is! (which is why you see it over at node/add)

**Comment by Graham-72** [comment]
>  If we were to build a view

I think having such a view would help us to compare the usability of the three approaches, and to understand this trial site. I could have a go at doing this if it will help?

**Comment by jenlampton** [comment]
Sure, just clone the other and change the node type filter. Do you have
access? Lmk if not.

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by Graham-72** [comment]
> Do you have access? Lmk if not.

Under Structure I have access to Taxonomy but not to Views at the moment.

**Comment by serundeputy** [comment]
> That's right. QA forum does not provide it's own page like forum module does (unless I am mistaken @serundeputy ?).

* QA Forum does not provide an admin page
* To see a 'view' of the question and answers you go to `/question/%nid` where `%nid` = the node id of the question.

**Comment by jenlampton** [comment]
> Compare / Contrast forum module options · Issue #3 · backdrop-ops/forum.backdropcms.org

not an admin page, a public listing like https://forum.backdropcms.org/forum-list

> To see a 'view' of the question and answers

Would this be the equivalent of the `node/%nid` page for forum topics?

> Under Structure I have access to Taxonomy but not to Views at the moment.

@Graham-72 you are now an administrator

**Comment by Graham-72** [comment]
I have added a view that lists the questions in the questions and answers type forum and linked it from a new main menu item 'Forum(QA)'

I would like to be able to include in the table a column 'Answer' but this data does not seem to be accessible from a view?

**Comment by serundeputy** [comment]
> I would like to be able to include in the table a column 'Answer' but this data does not seem to be accessible from a view?

Correct this data would not be accessible in a view since one of the ideas behind `qa_forum` is that making answers content types would require too many joins to be performant under load if there are more than 17 concurrent users (I made up the number 17), but performance could easily spin out of control with enough users posting several answers to many questions.

We could just put this problem off until it becomes a problem but if we choose an architecture that leads us right into the belly of the beast it may be a harder problem to solve.

**Comment by jenlampton** [comment]
I think we should stick with a simpler architecture that's already been proven, like nodes + comments. If/when we run into performance issues, we can first celebrate, and then devise a solution that best fits the problem we are having at that moment.

**Comment by FuzboleroXV** [comment]
FYI - I have started outlining the needs and considerations related to our "CMX qommunity distro" over here: https://cmx.arbolis.net/content/articles/qommunity-features-strategy-comments-everywhere%2C-so-'forum'-is-not-the-same-anymore
Will get several updates the coming days. Mainly we share @JenLampton's initial approach using core features instead of a dedicated forum module - plus it is important to be able to branch off comments into their own new posts, or even just move comments to another, existing post (from anywhere on the site, not necessarily "only" within forum context), hence our interest in the D7 "comment_mover" module's functionality: https://www.drupal.org/project/comment_mover

**Comment by klonos** [comment]
@FuzboleroXV can you please post that bullet list to [a new issue](https://github.com/backdrop-ops/forum.backdropcms.org/issues/new) so that we can take a look at it. If these are on GitHub instead of your own site, people can "watch" (follow) it for any updates and they can comment without having to create an account on a 3rd party site.

I have not read the article yet (due to lack of time), but I feel that there is a mix of issues that need to be branched off for core consideration, some for contrib and some that might belong here. Can't be sure though until there is a new issue created, so that should be step 1.

**Comment by docwilmot** [comment]
I think we've decided. Views and Comments it is!
